### DISCUSSION TITLE
Fields default values ignored after migrating from Babel to SWC

### DISCUSSION THREAD

**Discussion by fromi** [MichalLytek/type-graphql#1694] [Q&A]
I use typegraphql in a Next.js project. I am trying to move from Babel to SWC to reduce compile time, and everything works fine except for one thing: typegraphql used to return my default field value with babel when it was undefined in my return statement, and now with SWC, it remains undefined (leading to "cannot return null for non-nullable field" errors).

Here is an example:
```
@ObjectType()
export class GameOptions {
  @Field(() => Int)
  maxSpareTime = MAX_SPARE_TIME
```
When I return a GameOptions objet without this field (not very clean, but it comes from mongoose directly), with Babel, TypeGraphQL was automatically returning MAX_SPARE_TIME. With SWC, it does not.

Here is my Babel configuration:
```
{
  "presets": [
    "next/babel"
  ],
  "plugins": [
    "babel-plugin-transform-typescript-metadata", [
      "@babel/plugin-transform-typescript", {
        "allowDeclareFields": true
      }
    ], [
      "@babel/plugin-proposal-decorators", {
        "legacy": true
      }
    ], [
      "@babel/plugin-proposal-class-properties", {
        "loose": true
      }
    ]
  ]
}
```

Do you know why the default class structure was used with Babel, and why it is now used with SWC?
Is there simple way to get property fallback with TypeGraphQL? `@Field({default: ...})` does not work for return types. I can do custom resolvers, but it is a but messy of course. The best solution would be a way to configure SWC to have the same behavior as I use to have with Babel of course, to prevent unexpected regressions...

**Comment by MichalLytek** [comment]
TypeGraphQL uses a trick to detect default values. It creates a new instance of the class and then reads the instance's property value.
I have no experience to tell what settings SWC needs but in the meantime you can use `{ defaultValue }` workaround.

**Reply by fromi** [reply]
Ok! For now I am using v1.1.1, and `defaultValue` make the field nullable. The workaround does not work because it changes the result on client side, I get null instead of the default value.
In version 2.0, it will work?

**Reply by fromi** [reply]
Hi @MichalLytek ,
SWC team pointed out this option of Typescript configuration: https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-7.html#the-usedefineforclassfields-flag-and-the-declare-property-modifier
However, setting it does not work for me (yet).

**Reply by MichalLytek** [reply]
Yes, AFAIR it was fixed recently and in 2.0 default should not make the field nullable

**Reply by fromi** [reply]
Hi @MichalLytek,
How exactly do you create a new instance to read the instance's property value?
With Babel, the code is generated this way:
```
class Player {
  constructor() {
    Object.defineProperty(this, "preferences", {
    configurable: true,
    enumerable: true,
    writable: true,
      value: {}
    });
  }
}
```
With SWC, I have this:
```
class Player {
    constructor() {
	this.preferences = {};
  }
}
```
Why is it making any difference? Can you give me the file and line where you create the class instance?

**Reply by MichalLytek** [reply]
`const inputInstance = new (inputType.target as any)();`

**Reply by MichalLytek** [reply]
BTW `{}` is not a valid default value for graphql.

**Reply by fromi** [reply]
Hi @MichalLytek , I created a minimal reproduction repository here: https://github.com/fromi/next-typegraphql-field-default
You can start it by running `next dev`, then run this query:
```
fetch('http://localhost:3000/api/graphql', {
  method: 'POST',
  body: JSON.stringify({
    query: `{
      test {
        someField
      }
    }`
  }),
  headers: {
    'Content-type': 'application/json'
  }
})
.then(res => res.json())
.then(console.log)
```

You will get this error: `Cannot return null for non-nullable field Test.someField`

My resolver is returning `{}`. The class has one field:
```
@ObjectType()
export class Test {
  @Field()
  someField: string = 'hello'
}
```
When I was using Babel, typegraphql was automatically filling the missing property with the default value, on this repo it does not.
What is the expected behavior? The one with Babel or with SWC?

**Reply by fromi** [reply]
Hi @MichalLytek,
Still looking after the best solution for my issue. I would like to know what result is expected when you do that in 2.0:
```
@ObjectType()
export class Test {
  @Field({ defaultValue: 'hello' })
  someField: string = 'hello'
}

@Resolver(() => Test)
export class TestResolver {

  @Query(() => Test)
  async test() {
    return {} // 'hello' or 'cannot return null for non nullable field'?
  }
}
```
I would like to know what is the expected behavior (so I can fill a bug or not), right now I have to create a custom resolver to get fallback values for API retro-compatibility.

**Reply by MichalLytek** [reply]
![image](https://github.com/MichalLytek/type-graphql/assets/10618781/6d8d87b4-3f83-4285-8c13-ab7970965130)

https://typegraphql.com/docs/resolvers.html#arguments

**Reply by MichalLytek** [reply]
In other words: `defaultValue` is a schema SDL construct. We can set that via decorator options. It shows in the GraphQL schema then. But specification only allows for input and args to have default values, so it's ignored for object type and interfaces.

However, you can use some TS/JS approaches to create an object with default values. The initializer for property should work but your compiler does something weird and I don't know how to fix the compiler behavior.

**Reply by fromi** [reply]
I am not sure there is something wrong with the compiler. I investigated a little bit with help from SWC developers, and the javascript generated class looks like this:
```
class Test {
    constructor(){
        this.someField = "hello";
    }
}
```
Simple and straightforward. I tried to run `new Test()['someField']` to reproduce the behavior from TypeGraphQL, and I get the value as expected. There is something else done in TypeGraphQL that prevents the initializer for property to work when the property is not defined with [Object.defineProperty](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-7.html#the-usedefineforclassfields-flag-and-the-declare-property-modifier).

**Reply by MichalLytek** [reply]
https://github.com/carlocorradini/type-graphql/blob/master/src/helpers/types.ts#L87-L110

**Reply by MichalLytek** [reply]
this is the code used for output types like object type

**Reply by MichalLytek** [reply]
but it's not used to transform value `{}` from returned from resolver, only to parse it as `@Root()` in field resolvers

**Reply by MichalLytek** [reply]
in other words, please create a `@FieldResolver()` and return there the default value or `root.foobar` if defined or new api or something else

**Reply by fromi** [reply]
Ok, thanks! So, the fact that it worked with Babel was probably some sort of side effect.
I already created a FieldResolver for the most problematic point in our API, the rest was just made nullable to make it work.
It is a bit sad though, having to create a FieldResolver to create a fallback for undefined values is way more verbose.

Thank you for your help
